File: src/feats/dag.rs
//! FeatureDag facade —— 委派给 DagBuilder / DagExecutor / FeatureInfo。
//!
//! 新代码请直接使用 builder/executor/feature_info 模块，而非此 facade。

pub use super::builder::{DagArtifact, DagBuilder, ValidationIssue, ValidationReport};
pub use super::executor::{DagExecutor, ExecStep, ExecutionPlan};
use super::feature_info::{FeatureInfo, FeatureScope};
use super::ops::Fv;
use crate::feats::config::{DataSourceDef, FlowConfig, OperatorDef, SourceDef};
use crate::feats::debug::DebugTracer;
use crate::feats::defaults::source_default;
use crate::feats::schema::FeatureSchema;
use std::collections::{HashMap, HashSet};

/// 特征值的类型别名。
pub type FeatureValue = Fv;

/// 特征处理结果：包含所有特征值并区分来源。
pub struct FeatureResult {
    /// 所有 source 与计算结果的特征值。
    pub features: HashMap<String, FeatureValue>,
    /// 原始 source 名称集合。
    pub source_names: HashSet<String>,
    /// DAG 计算产生的特征名称集合。
    pub computed_names: HashSet<String>,
}

/// 特征 DAG 执行引擎（facade）。
///
/// 根据 FlowConfig 构建算子图，拓扑排序后按序执行单样本处理。
pub struct FeatureDag {
    sources: HashMap<String, SourceDef>,
    /// 配置中的外部数据源定义。
    pub data_sources: Vec<DataSourceDef>,
    node_defs: HashMap<String, OperatorDef>,
    /// 拓扑排序后的算子执行顺序。
    pub execution_order: Vec<String>,
    debug_mode: bool,
    /// 可选调试 tracer，用于记录执行过程。
    pub tracer: Option<DebugTracer>,
    /// 预编译执行计划。
    pub plan: ExecutionPlan,
    /// source 和算子输出的静态 schema。
    pub feature_schemas: HashMap<String, FeatureSchema>,
    /// DAG 构建校验报告。
    pub validation_report: ValidationReport,
}

impl FeatureDag {
    /// 从 FlowConfig 构建 FeatureDag，包含拓扑排序和预编译执行计划。
    pub fn from_config(
        config: FlowConfig,
        debug_mode: bool,
        tracer: Option<DebugTracer>,
    ) -> Result<Self, String> {
        let artifact = DagBuilder::build(config)?;
        let data_sources = artifact.data_sources;
        let sources = artifact.sources;
        let node_defs = artifact.node_defs;
        let execution_order = artifact.execution_order;
        let plan = artifact.plan;
        let feature_schemas = artifact.feature_schemas;
        let validation_report = artifact.validation_report;

        Ok(Self {
            sources,
            data_sources,
            node_defs,
            execution_order,
            debug_mode,
            tracer,
            plan,
            feature_schemas,
            validation_report,
        })
    }

    /// 获取原始输入源定义。
    pub fn source_defs(&self) -> &HashMap<String, SourceDef> {
        &self.sources
    }

    /// 获取算子节点定义。
    pub fn operator_defs(&self) -> &HashMap<String, OperatorDef> {
        &self.node_defs
    }

    /// 获取所有需要嵌入的特征 (name, EmbedConfig) 列表。
    pub fn embeddable_features(&self) -> Vec<(&str, &crate::feats::config::EmbedConfig)> {
        let mut result = Vec::new();
        for (_, op) in &self.node_defs {
            if let Some(ref emb) = op.embed {
                for out_name in &op.outputs {
                    result.push((out_name.as_str(), emb));
                }
            }
        }
        result.sort_by(|a, b| a.0.cmp(b.0));
        result
    }

    /// Classify each operator by user/item/context-derived feature scope.
    pub fn op_source_kind(&self) -> HashMap<String, FeatureScope> {
        FeatureInfo::new(
            self.sources.clone(),
            self.node_defs.clone(),
            self.feature_schemas.clone(),
            self.execution_order.clone(),
        )
        .op_source_kind()
    }

    /// Get source names (used to filter inputs in broadcast mode).
    pub fn source_names(&self) -> Vec<&str> {
        self.sources.keys().map(|s| s.as_str()).collect()
    }

    /// Get operator outputs by operator name.
    pub fn op_outputs(&self, op_name: &str) -> Option<&Vec<String>> {
        self.node_defs.get(op_name).map(|d| &d.outputs)
    }

    /// 执行 DAG：raw_inputs → 拓扑序执行算子 → FeatureResult。
    pub fn execute(
        &self,
        raw_inputs: &HashMap<String, FeatureValue>,
    ) -> Result<FeatureResult, String> {
        if let Some(ref tracer) = self.tracer {
            tracer.begin_sample();
        }

        let mut context: HashMap<String, FeatureValue> = HashMap::new();

        let mut overridden = Vec::new();
        for (name, val) in raw_inputs {
            if self.sources.contains_key(name) {
                context.insert(name.clone(), val.clone());
                overridden.push(name.clone());
            }
        }

        for (name, source_def) in &self.sources {
            if !context.contains_key(name) {
                let default_val = source_default(source_def);
                context.insert(name.clone(), default_val);
            }
        }

        let _ = &self.tracer;

        let source_names: HashSet<String> = self.sources.keys().cloned().collect();
        let mut computed_names = HashSet::new();
        for (op_idx, node_name) in self.execution_order.iter().enumerate() {
            let op = &self.plan.ops[op_idx];
            let def = &self.node_defs[node_name];
            let op_inputs: Vec<Fv> = def
                .inputs
                .iter()
                .map(|inp| {
                    context
                        .get(inp)
                        .map(|v| v.clone())
                        .ok_or_else(|| format!("Required input '{}' not found", inp))
                })
                .collect::<Result<_, String>>()?;
            let output = op.process(&op_inputs)?;
            let output_fv: FeatureValue = output;
            for out_name in &def.outputs {
                context.insert(out_name.clone(), output_fv.clone());
                computed_names.insert(out_name.clone());
            }
        }
        if self.debug_mode {
            self.dump_snapshot(&context, &source_names, &computed_names);
        }
        if let Some(ref tracer) = self.tracer {
            tracer.end_sample();
        }
        Ok(FeatureResult {
            features: context,
            source_names,
            computed_names,
        })
    }

    /// Batch DAG execution on columnar data.
    pub fn execute_batch(
        &self,
        columns: &HashMap<String, Vec<FeatureValue>>,
        skip_ops: &HashSet<String>,
    ) -> Result<HashMap<String, Vec<FeatureValue>>, String> {
        self.execute_batch_precomputed(columns, skip_ops, &HashMap::new())
    }

    /// Batch execution with precomputed single-value features broadcast to all rows.
    pub fn execute_batch_precomputed(
        &self,
        columns: &HashMap<String, Vec<FeatureValue>>,
        skip_ops: &HashSet<String>,
        precomputed: &HashMap<String, FeatureValue>,
    ) -> Result<HashMap<String, Vec<FeatureValue>>, String> {
        let n_rows = columns.values().next().map(|v| v.len()).unwrap_or(0);
        if n_rows == 0 {
            return Ok(HashMap::new());
        }

        let mut context: HashMap<String, Vec<FeatureValue>> = HashMap::new();

        for (name, col) in columns {
            if self.sources.contains_key(name) {
                context.insert(name.clone(), col.clone());
            }
        }
        for (name, source_def) in &self.sources {
            if !context.contains_key(name) {
                let default = source_default(source_def);
                context.insert(name.clone(), vec![default; n_rows]);
            }
        }
        for (name, val) in precomputed {
            context.insert(name.clone(), vec![val.clone(); n_rows]);
        }

        let default_col: Vec<FeatureValue> = vec![Fv::Int(0); n_rows];
        for (op_idx, node_name) in self.execution_order.iter().enumerate() {
            if skip_ops.contains(node_name) {
                continue;
            }
            let op = &self.plan.ops[op_idx];
            let def = &self.node_defs[node_name];

            let input_slices: Vec<&[FeatureValue]> = def
                .inputs
                .iter()
                .map(|inp| {
                    context
                        .get(inp)
                        .map(|c| c.as_slice())
                        .unwrap_or(default_col.as_slice())
                })
                .collect();

            let result_vec = op
                .process_batch(&input_slices, n_rows)
                .map_err(|e| format!("{}: {}", node_name, e))?;

            for out_name in &def.outputs {
                context.insert(out_name.clone(), result_vec.clone());
            }
        }

        Ok(context)
    }

    /// 调试用：输出当前 context 的快照日志。
    pub fn dump_snapshot(
        &self,
        context: &HashMap<String, FeatureValue>,
        source_names: &HashSet<String>,
        computed_names: &HashSet<String>,
    ) {
        tracing::info!("Feature Snapshot");
        for (name, val) in context {
            let type_name = val.type_name();
            let origin = if computed_names.contains(name) {
                "computed"
            } else if source_names.contains(name) {
                "source"
            } else {
                "raw"
            };
            tracing::info!(origin, name, type_name, "feature snapshot entry");
        }
    }
}

File: python/src/train/core/dag.py
from __future__ import annotations

"""特征 DAG 执行器 —  facade，委派给 builder / executor / feature_info / preprocessor。

新代码请直接使用 builder、executor、feature_info、preprocessor 模块，而非此 facade。
"""

import logging
from dataclasses import dataclass
from typing import Any

import torch

from ..ops import CustomOp, create_op
from .builder import (
    DagBuilder,
    ValidationReport,
    parse_default,
)

_build = DagBuilder.build
from .config import EmbedConfig, FlowConfig, OperatorDef, SourceDef
from .executor import DagExecutor
from .feature_info import FeatureInfo
from .preprocessor import DagPreprocessor
from .schema import FeatureSchema

logger = logging.getLogger(__name__)

FeatureValue = Any


@dataclass
class FeatureResult:
    features: dict[str, FeatureValue]
    source_names: set[str]
    computed_names: set[str]


class FeatureDag:
    def __init__(
        self,
        config: FlowConfig,
        debug_mode: bool = False,
        tracer: Any | None = None,
        strict_validation: bool = False,
    ) -> None:
        artifact = _build(config)
        self._artifact = artifact
        self._executor = DagExecutor(
            artifact.plan,
            artifact.sources,
            node_defs=artifact.node_defs,
            execution_order=artifact.execution_order,
        )
        self._feat_info = FeatureInfo(
            artifact.sources,
            artifact.node_defs,
            artifact.feature_schemas,
            artifact.execution_order,
        )
        self._preprocessor = DagPreprocessor(self._feat_info)

        self.sources: dict[str, SourceDef] = artifact.sources
        self.nodes: dict[str, CustomOp] = {}
        op_name_to_idx: dict[str, int] = {}
        for i, node_name in enumerate(artifact.execution_order):
            op = artifact.plan._ops[i]
            self.nodes[node_name] = op
            op_name_to_idx[node_name] = i
        self.node_defs: dict[str, OperatorDef] = artifact.node_defs
        self.execution_order: list[str] = artifact.execution_order
        self.debug_mode = debug_mode
        self.tracer = tracer
        self.feature_schemas: dict[str, FeatureSchema] = artifact.feature_schemas
        self.validation_report: ValidationReport = artifact.validation_report
        self._source_names = tuple(artifact.sources)
        self._source_name_set = set(artifact.sources)
        self._embed_infos = dict(self._feat_info.embeddable_features())
        self._embed_names = tuple(self._embed_infos)

        if strict_validation and self.validation_report.warnings:
            details = ", ".join(issue.message for issue in self.validation_report.warnings)
            raise ValueError(f"strict validation failed: {details}")

    def _validate(self, config: FlowConfig) -> ValidationReport:
        return self.validation_report

    @property
    def executor(self) -> DagExecutor:
        return self._executor

    @property
    def feat_info(self) -> FeatureInfo:
        return self._feat_info

    @property
    def preprocessor(self) -> DagPreprocessor:
        return self._preprocessor

    @staticmethod
    def _parse_default(val_str: str, dtype: Any) -> FeatureValue:
        return parse_default(val_str, dtype)

    @staticmethod
    def _create_op(def_: OperatorDef) -> CustomOp:
        return create_op(def_.op_type, def_.params)

    def embeddable_features(self) -> list[tuple[str, EmbedConfig]]:
        return self._feat_info.embeddable_features()

    def feature_tuples(self) -> list[tuple[str, int, int]]:
        return self._feat_info.feature_tuples()

    def feature_total_dim(self) -> int:
        return self._feat_info.feature_total_dim()

    def feature_pooling(self) -> dict[str, str]:
        return self._feat_info.feature_pooling()

    def feature_seq_lens(self) -> dict[str, int]:
        return self._feat_info.feature_seq_lens()

    def execute_batch(self, columns: dict[str, list]) -> dict[str, list]:
        n_rows = len(next(iter(columns.values()))) if columns else 0
        if n_rows == 0:
            return {}

        context: dict[str, list] = {}
        for name, col in columns.items():
            if name in self.sources:
                context[name] = list(col)

        for name, src in self.sources.items():
            if name not in context:
                default = self._parse_default(src.default_val, src.dtype)
                context[name] = [default] * n_rows
            else:
                col = context[name]
                if any(v is None for v in col):
                    default = self._parse_default(src.default_val, src.dtype)
                    context[name] = [default if v is None else v for v in col]

        for node_name in self.execution_order:
            op, def_ = self.nodes[node_name], self.node_defs[node_name]
            op_inputs = [context[inp] for inp in def_.inputs]

            if hasattr(op, "process_batch"):
                output = op.process_batch(op_inputs)
            else:
                output = []
                for i in range(n_rows):
                    row_inputs = [col[i] for col in op_inputs]
                    output.append(op.process(row_inputs))

            if def_.outputs:
                for out_name in def_.outputs:
                    context[out_name] = output

        return context

    def preprocess_batch(self, rows: list[dict] | dict[str, list]) -> dict[str, torch.Tensor]:
        if isinstance(rows, dict):
            columns = {
                name: list(values) for name, values in rows.items() if name in self._source_name_set
            }
        else:
            n_rows = len(rows)
            if n_rows > 0:
                import pandas as pd
                df = pd.DataFrame(rows)
                columns = {
                    col: df[col].tolist() for col in df.columns if col in self._source_name_set
                }
            else:
                columns = {}

        if self.tracer:
            n_rows = len(next(iter(columns.values()))) if columns else 0
            for i in range(n_rows):
                self.tracer.begin_sample(i)

        result = self.execute_batch(columns)

        if self.tracer:
            n_rows = len(next(iter(columns.values()))) if columns else 0
            for _ in range(n_rows):
                self.tracer.end_sample()

        return self._preprocessor.preprocess(result)

    def execute(self, raw_inputs: dict[str, FeatureValue], sample_id: int = 0) -> FeatureResult:
        context: dict[str, FeatureValue] = {
            name: val for name, val in raw_inputs.items() if name in self.sources
        }

        overridden = list(raw_inputs.keys())
        for name, src in self.sources.items():
            if name not in context:
                context[name] = self._parse_default(src.default_val, src.dtype)

        if self.tracer:
            self.tracer.trace_defaults(context)
            self.tracer.trace_overrides(context, overridden)

        source_names = set(self.sources.keys())
        computed_names: set[str] = set()

        for node_name in self.execution_order:
            op, def_ = self.nodes[node_name], self.node_defs[node_name]
            op_inputs = [context[inp] for inp in def_.inputs]
            output = op.process(op_inputs)
            if self.tracer:
                self.tracer.trace_operator(node_name, def_.inputs, op_inputs, def_.outputs, output)
            for out_name in def_.outputs:
                context[out_name] = output
                computed_names.add(out_name)

        if self.debug_mode:
            self._dump_snapshot(context, source_names, computed_names)
        if self.tracer:
            self.tracer.end_sample()
        return FeatureResult(
            features=context,
            source_names=source_names,
            computed_names=computed_names,
        )

    def _dump_snapshot(
        self,
        context: dict[str, FeatureValue],
        source_names: set[str],
        computed_names: set[str],
    ) -> None:
        logger.debug("[Feature Snapshot]")
        for name, val in sorted(context.items()):
            origin = (
                "computed"
                if name in computed_names
                else "source"
                if name in source_names
                else "raw"
            )
            logger.debug(" -> [%s] %-20s | value=%s", origin, name, val)
